File: frontend/src/api/sale.ts
import apiClient from './client';

export interface Sale {
  id: string;
  factory_id?: string;
  date: string;
  customer_name: string;
  customer_phone?: string;
  product_id: string;
  quantity_sold: number;
  rate_per_brick: number;
  total_amount: number;
  amount_received: number;
  balance_due: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface SaleCreate {
  factory_id: string;
  date: string;
  customer_name: string;
  customer_phone?: string;
  product_id: string;
  quantity_sold: number;
  rate_per_brick: number;
  total_amount: number;
  amount_received?: number;
  balance_due?: number;
  notes?: string;
}

export const saleApi = {
  getByFactory: async (factoryId: string, startDate?: string, endDate?: string): Promise<Sale[]> => {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);
    const response = await apiClient.get(`/sales/factory/${factoryId}?${params.toString()}`);
    return response.data;
  },

  getById: async (id: string): Promise<Sale> => {
    const response = await apiClient.get(`/sales/${id}`);
    return response.data;
  },

  create: async (data: SaleCreate): Promise<Sale> => {
    const response = await apiClient.post('/sales', data);
    return response.data;
  },

  update: async (id: string, data: Partial<Sale>): Promise<Sale> => {
    const response = await apiClient.put(`/sales/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/sales/${id}`);
  },
};

File: backend/models/sale.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime, date

class SaleBase(BaseModel):
    date: date
    customer_name: str
    customer_phone: Optional[str] = None
    product_id: str
    quantity_sold: int
    rate_per_brick: float
    total_amount: float
    amount_received: float = 0.0
    balance_due: float = 0.0
    notes: Optional[str] = None

class SaleCreate(SaleBase):
    factory_id: str

class SaleUpdate(BaseModel):
    date: Optional[date] = None
    customer_name: Optional[str] = None
    customer_phone: Optional[str] = None
    product_id: Optional[str] = None
    quantity_sold: Optional[int] = None
    rate_per_brick: Optional[float] = None
    total_amount: Optional[float] = None
    amount_received: Optional[float] = None
    balance_due: Optional[float] = None
    notes: Optional[str] = None

class Sale(SaleBase):
    id: str = Field(default_factory=lambda: str(__import__('uuid').uuid4()))
    factory_id: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Config:
        from_attributes = True